### PR TITLE
Polarity-based Splitting Heuristics

### PR DESCRIPTION
**Pull request by anwu1219**
Add polarity-based splitting heuristics as one of the internal splitting strategy. 
On 180 ACAS benchmarks with 20 min timeout, new strategy solved 116 instance in 13390 secs. Original splitting strategy solved 116 instances in 17142 secs.

Still need to add/fix some tests.

### CODE REVIEW

**Comment by ahmed-irfan** [comment]
is it ready for review? it is tagged with `do-no-merge` label...

**Comment by anwu1219** [comment]
> is it ready for review? it is tagged with `do-not-merge` label...

Not yet. Still need to fix something. Will ping when it's ready.

File: src/engine/SmtCore.cpp [hunk 1]
```diff
@@ -62,10 +62,9 @@ void SmtCore::reportViolatedConstraint( PiecewiseLinearConstraint *constraint )
          _constraintViolationThreshold )
     {
         _needToSplit = true;
-        if ( GlobalConfiguration::SPLITTING_HEURISTICS == DivideStrategy::ReLUViolation )
+        if ( GlobalConfiguration::SPLITTING_HEURISTICS ==
+             DivideStrategy::ReLUViolation || !pickSplitPLConstraint() )
             _constraintForSplitting = constraint;
-        else
```
Comment (ahmed-irfan) [inline src/engine/SmtCore.cpp:67]: no else needed ?

File: src/engine/Engine.cpp [hunk 2]
```diff
@@ -1920,23 +1920,68 @@ void Engine::updateDirections()
...
+            unsigned size = _networkLevelReasoner->getLayerSize( layer );
+            for ( unsigned neuron = 0; neuron < size; ++neuron ){
+                auto plConstraint = _networkLevelReasoner->
+                    getPLConstraintFromIndex( layer, neuron );
+                if ( plConstraint && plConstraint->isActive()
+                     && !plConstraint->phaseFixed() )
+                {
+                    plConstraint->updateScore();
+                    _candidatePlConstraints.insert( plConstraint );
+                    if ( _candidatePlConstraints.size() >=
```
Comment (ahmed-irfan) [inline src/engine/Engine.cpp:1943]: I would suggest putting this check in the outer for loop and removing the return statements.

File: src/engine/SmtCore.cpp (hunk 1 above)
Comment (anwu1219) [reply src/engine/SmtCore.cpp:67]: Here I changed it such that if the alternative heuristic fail to pick a splitting ReLU (e.g., when NetworkLevelReasoner is not available), we fail back to the native ReluViolation count based  strategy.

Comment (anwu1219) [reply src/engine/SmtCore.cpp:67]: added a comment

File: src/engine/Engine.cpp [hunk 3]
```diff
@@ -1920,23 +1920,69 @@ void Engine::updateDirections()
...
+                         GlobalConfiguration::RUNTIME_ESTIMATE_THRESHOLD )
+                        break;
+                }
+            }
         }
+        return;
```
Comment (ahmed-irfan) [inline src/engine/Engine.cpp:1950]: remove this return

File: src/engine/Engine.cpp [hunk 4]
```diff
@@ -1920,23 +1920,69 @@ void Engine::updateDirections()
...
+    {
+        for ( const auto plConstraint : _plConstraints )
+        {
+            if ( plConstraint->isActive() && !plConstraint->phaseFixed() )
+            {
+                plConstraint->updateScore();
+                _candidatePlConstraints.insert( plConstraint );
+            }
+        }
+        return;
```
Comment (ahmed-irfan) [inline src/engine/Engine.cpp:1963]: remove this return

File: src/engine/Engine.cpp (hunk 2 above)
Comment (anwu1219) [reply src/engine/Engine.cpp:1944]: Made the change accordingly.

**Pull request merged**